proxy/mempool/mempool_schedule.py

```diff
@@ -406,7 +477,7 @@ def add_tx(self, tx: MPTxRequest) -> MPTxSendResult:
             return MPTxSendResult(code=MPTxSendResultCode.Underprice, state_tx_cnt=None)
 
         if self.tx_cnt >= self._capacity:
-            lower_tx = self._tx_dict.peek_lower_tx()
+            lower_tx = self._tx_dict.peek_gapped_lower_tx()
```

Comment:
Right now,` peek_gapped_lower_tx()` returns the gapped tx with the highest gas-price.
As a result, it is impossible to put any tx with the normal gas-price to the mempool.

For the full mempool the logic should be:
1. If the tx is gapped
1.a.  Check that `lowest-gapped-tx.gas-price` < `tx.gas-price`
1.b.  If no gapped-txs -> reject the tx
2. If the tx is pending
2.a. If the mempool has gapped txs -> success -> accept tx, because it will remove a gapped tx
2.b. If the mempool doesn't have gapped txs -> check the `lowest-pending-tx.gas-price < tx.gas-price`

Reply:
Split them out? Ah, nice. Ok sounds good.

Reply:
Just thinking about `2.b` - that would mean that a single tx with a high gas price can evict a lower gas `pending` tx. This is fine... but I'm thinking what if this soon-to-be-evicted-tx is at the start of a sequence of pending transactions? It would mean that one transaction could knock out a whole sequence of pending transactions.

I've just looked at Geth and it seems that they do this as well.

If we're concerned in the future, what we could do is add up all the gas prices for the whole sequence and only evict if that single tx gas price > total-soon-to-be-evicted-tx gas price

Reply:
Thinking about this more...

> only evict if that single tx gas price > total-soon-to-be-evicted-tx gas price

Adding this single high gas price tx might actually move txs from gapped to non-gapped, so in future the comparison should be "(single-tx + all-gapped-to-gas-moved-tx caused by adding single-tx) > total-soon-to-be-evicted-tx"

Reply:
Ok.
Add this task to the jira.
Let's disuccuss it later